src/uct/ib/base/ib_iface.c

```diff
@@ -1173,6 +1174,131 @@ static void uct_ib_iface_set_path_mtu(uct_ib_iface_t *iface,
     }
 }
 
+const char*
+uct_ib_iface_sl_mask_str(uint16_t sl_mask, ucs_string_buffer_t *str_b)
+{
+    uint8_t sl;
+
+    if (sl_mask == 0) {
+        ucs_string_buffer_appendf(str_b, "<none>");
+        goto out;
+    }
+
+    ucs_for_each_bit(sl, sl_mask) {
+        ucs_assert(sl < 16);
+
+        if (ucs_string_buffer_length(str_b) == 0) {
+            ucs_string_buffer_appendf(str_b, "%u", sl);
+        } else {
+            ucs_string_buffer_appendf(str_b, ", %u", sl);
+        }
+    }
+
+out:
+    return ucs_string_buffer_cstr(str_b);
+}
+
+static ucs_status_t
+uct_ib_iface_select_sl_result(const uct_ib_iface_config_t *ib_config,
+                              ucs_ternary_value_t ar_enable,
+                              uint16_t ooo_sl_mask,
+                              int ooo_sl_mask_is_detected,
+                              ucs_string_buffer_t *sl_str,
+                              const char *dev_port_name)
+{
+    if (ar_enable == UCS_YES) {
+        if (ooo_sl_mask_is_detected) {
```

Comment:
imo this handling is overcomplicated.
Maybe better return the previous implementation and just the error string?
So you just warn user that AR may not be supported if you can't detect it?

Reply:
fixed